Add rtl_433 Prometheus exporter

zopieux opened the pull request:
###### Motivation for this change

NixOS already packages rtl_433, an SDR receiver that among other things can decode metrics broadcast by 433MHz appliances such as thermometers.

This PR adds a lightweight Prometheus exporter written in Go for rtl_433 metrics.

I had to add further pkgs/modules (see first commits) to make that work seamlessly:
1. upstream rtl-sdr (0.6.0) is using a very permissive udev ruleset that makes devices o+rw, which is not okay, so I backport a commit from their master branch to make it more locked-down
1. because of 1. I had to make sure the "plugdev" group exists, so I introduce a small hardware.rtl-sdr module that enables said udev rules and ensures the group exists
1. finally, I add the rtl_433_prometheus package and exporter module
 
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

delroth commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,25 @@
+
+  vendorSha256 = "03mnmzq72844hzyw7iq5g4gm1ihpqkg4i9dgj2yln1ghwk843hq6";
+
+  doCheck = false;
```
Could you add a comment explaining why the tests are disabled? (or enable them)

delroth commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,25 @@
+  meta = with lib; {
+    description = "Prometheus time-series DB exporter for rtl_433 433MHz radio packet decoder";
+    homepage = "https://github.com/mhansen/rtl_433_prometheus";
+    license = licenses.mit;
```
maintainers? If you're not already in the list, please add yourself (in a separate commit).

delroth commented on nixos/modules/services/monitoring/prometheus/exporters/rtl_433.nix:
```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, options }:
+
+with lib;
```
I'm in general not a big fan of "with", especially for something as broad as "lib". In some cases like for `meta = with lib; {` it's fairly well accepted but here it's not actually saving much imo.

Zopieux replied:
Sure, removed.

WilliButz replied:
for future reference (but not relevant for this pr):
instead of using `with lib;` to pull everything from `lib` in scope or specifying `lib.` everywhere, one can also pull in only the needed defintions as follows:
```nix
{ lib, ... }:
let
  inherit (lib)
    mkOption
    # ... more stuff from lib
    types;
in
{
  options.example-module-option = mkOption {
    type = types.bool;
    # ...
  };
  # ...
}
```

delroth commented on nixos/modules/services/monitoring/prometheus/exporters/rtl_433.nix:
```diff
@@ -0,0 +1,84 @@
+{ config, lib, pkgs, options }:
```
Why is "options" in here? AFAICT it's not used anywhere.

Zopieux replied:
It's not used but it's [passed by Prometheus import](https://github.com/NixOS/nixpkgs/blob/97005a006c7a62aebdd79ccc2b2eb179991465bd/nixos/modules/services/monitoring/prometheus/exporters.nix#L53). Removing it fails the build.

delroth commented on nixos/modules/services/monitoring/prometheus/exporters/rtl_433.nix:
```diff
@@ -0,0 +1,84 @@
+    };
+  in
+  {
+    flags = mkOption {
```
I would rename to e.g. rtl433Flags to avoid confusion with extraFlags.

Zopieux replied:
Done.

delroth reviewed (changes requested):
Please check https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches re: commit message formatting.

In general it's also preferred to split off package and NixOS module into two separate commits.

zopieux committed:
maintainers: add zopieux

zopieux committed:
prometheus-rtl_433-exporter: init at 641a1f5

Zopieux commented:
Thanks for the quick review. PTAL, refactored how I build the matcher options in the module.

delroth commented on nixos/modules/services/monitoring/prometheus/exporters/rtl_433.nix:
```diff
@@ -0,0 +1,75 @@
+      default = "";
+      example = "-R 19";
+      description = ''
+        Flags passed verbatim to rtl_433_prometheus binary.
```
That seems incorrect -- the flags are passed to rtl_433, not rtl_433_prometheus.

Zopieux replied:
Good catch, fixed.

delroth reviewed (approved):
Just one small issue left.

zopieux committed:
nixos/prometheus-rtl_433-exporter: new module

zopieux committed:
rtl-sdr: use hardened udev rules

rtl-sdr latest release 0.6.0 is from 2018. Six months ago master received
a patch[1] to use improved udev rules that chown to plugdev instead of
making the devices world-read-writable.

This change backports [1] as an upstream patch.

[1] https://osmocom.org/projects/rtl-sdr/repository/revisions/b2814731563be4d5a0a68554ece6454a2c63af12

zopieux committed:
nixos/hardware/rtl-sdr: new module

This is a very thin module to ensure "plugdev" user exists and some udev
rules are enabled, a prerequisite for using rtl-sdr.

Zopieux commented on nixos/modules/services/monitoring/prometheus/exporters.nix:
```diff
@@ -224,6 +225,8 @@ in
     services.prometheus.exporters.minio.minioAccessSecret = mkDefault config.services.minio.secretKey;
   })] ++ [(mkIf config.services.rspamd.enable {
     services.prometheus.exporters.rspamd.url = mkDefault "http://localhost:11334/stat";
+  })] ++ [(mkIf config.services.prometheus.exporters.rtl_433.enable {
+    hardware.rtl-sdr.enable = mkDefault true;
```
@willibutz Hi, I'm introducing a prom exporter that has a dependency on {a udev ruleset, the `plugdev` group}, which I've abstracted away through this `config.hardware.rtl-sdr` module (available in this PR).

I'm trying to figure out the proper way to enable the `hardware.rtl-sdr` module when the `rtl_433` exporter is enabled too. The surrounding lines are doing the opposite (changing some exporter options when an external module is enabled) so I'm a bit confused what to do – I'm by no mean a NixOS expert. Thanks for your help!

WilliButz replied:
There is currently no nice way to do this, I'll need to take another look at moving these defaults in the exporter submodules. So I'd say this is fine, and better than setting this default in the `harware.rtl-sdr` module.

Zopieux replied:
ACK, not touching this.

delroth commented on nixos/modules/hardware/rtl-sdr.nix:
```diff
@@ -0,0 +1,22 @@
+
+in {
+  options.hardware.rtl-sdr = {
+    enable = lib.mkEnableOption {
```
This should be lib.mkEnableOption '' ... ''; (it doesn't take an attrset)

Zopieux replied:
Done.

delroth commented:
FYI I'm now using this successfully on one of my systems. https://github.com/delroth/infra.delroth.net/commit/067c7092c2c4815c3fd0b64f84af662df5ba2bb9

@WilliButz any way you could help move this PR forward? I've done some review passes already and I feel like it's in a good shape, modulo one comment on "how to enable the hardware.* support when the exporter is enabled" still pending.

Zopieux commented:
Upstream [tagged a release](https://github.com/mhansen/rtl_433_prometheus/issues/6), updated to use it instead of commit hash.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/317

WilliButz commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,24 @@
+{ lib, buildGoModule, fetchFromGitHub, pkgs }:
```
```suggestion
{ lib, buildGoModule, fetchFromGitHub, bash, nixosTests }:
```

Zopieux replied:
Applied.

WilliButz commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,24 @@
+    sha256 = "1998gvfa5310bxhi6kfv8bn99369dxph3pwrpp335997b25lc2w2";
+  };
+
+  postPatch = "substituteInPlace rtl_433_prometheus.go --replace /bin/bash ${pkgs.bash}/bin/bash";
```
```suggestion
  postPatch = "substituteInPlace rtl_433_prometheus.go --replace /bin/bash ${bash}/bin/bash";
```

Zopieux replied:
Applied.

WilliButz commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,24 @@
+  postPatch = "substituteInPlace rtl_433_prometheus.go --replace /bin/bash ${pkgs.bash}/bin/bash";
+
+  vendorSha256 = "03mnmzq72844hzyw7iq5g4gm1ihpqkg4i9dgj2yln1ghwk843hq6";
+
```
```suggestion

passthru.tests = { inherit (nixosTests.prometheus-exporters) rtl_433; };

```

Zopieux replied:
Applied.

WilliButz commented on pkgs/servers/monitoring/prometheus/rtl_433-exporter.nix:
```diff
@@ -0,0 +1,24 @@
+{ lib, buildGoModule, fetchFromGitHub, pkgs }:
+
+buildGoModule rec {
+  name = "rtl_433-exporter";
```
```suggestion
  pname = "rtl_433-exporter";
```

WilliButz replied:
Maybe even rename the file and set pname to `prometheus-rtl_433-exporter`, since it is later referred to as that (starting with `all-packages.nix`). But as long as it is consistent in the end, I am fine with the current name as well :)

Zopieux replied:
Looking at neighboring files, they all use `pname` without the `prometheus-` suffix. Keeping it that way for consistency. Please reopen if you feel strongly about it :-)

WilliButz replied:
Oh right, that's fine :+1:

WilliButz reviewed (changes requested):
Please excuse my delayed review, I made a few small improvement suggestions, otherwise this looks good to me.

Thanks for the contribution :+1:

WilliButz reviewed (approved):
:+1:
